proposals/0000-type-vs-constraint.rst

```diff
@@ -0,0 +1,393 @@
+``axDefault`` is *heterogeneous*: the kind of ``Default a`` is ``Constraint``,
+whereas the kind of ``a`` is ``Type``.
+
+GHC allows us to extract out an equality relationship between *kinds* from an
+equality relationship on *types* -- and kind equalities are always nominal. To
+wit, Core allows ::
+
+  KindCo axDefault :: Constraint ~# Type
+
+Now, suppose that you could write this::
+
+  type family F a
+  type instance F Type = Int
+  type instance F Constraint = Bool
+
+If these instances were allowed, GHC could
+produce a coercion between ``Int`` and ``Bool``, thus::
+
+  Bool  ~#  F Constraint   -- By type instance F Constraint (backwards)
+        ~#  F Type         -- By KindCo axDefault
+        ~#  Int            -- By type instance F Type
+
+That would be Very, Very Bad.  So, although ``Type`` and ``Constraint`` are built
+with different (un-equal) primitive type constructors,
+
+* **GHC's type checker treats `Type` and `Constraint` as not "apart".**
+
+That in turn makes GHC complain that the above instances overlap, and are hence illegal.
+You can read more about what "apart" means in
+`Closed type families with overlapping equations <https://simon.peytonjones.org/closed-type-families/>`_,
+sections 3.2 and 3.3.
+
+Apartness affects type-class instances as well as type-family instances.
+Suppose we have::
+
+  instance {-# OVERLAPPABLE #-} C Int a  where ...
+  instance {-# OVERLAPPING #-}  C Int Bool where ...
+
+and we are trying to solve the constraint ``C Int (F gamma)``, where ``F`` is a type family and ``gamma``
+is an as-yet-unknown unification variable. Since ``(F gamma)`` is not
+apart from ``Bool``, it could be that (once we know ``gamma``) the second
+instance should be chosen. So GHC declines to commit to the first.
+
+If ``(F gamma)`` later simplifes to, say ``Char`` (which *is* apart from ``Bool``),
+then and only then GHC can commit to the first instance.
+
+Similarly with ``Type`` and ``Constraint``. Suppose we have::
+
+  instance D Type where...
+  instance D Constraint where ...
+
+and try to solve ``C Int Type``. This matches the first instance, but
+since ``Type`` is not apart from ``Constraint`` GHC thinks "oh, the
+second one could match in the future" and declines to commit to the
+first.  In fact, you can have either of these instances separately,
+but if they occur together neither will ever be chosen: they overlap
+irretrievably.
+
+
+Proposed Change Specification
+-----------------------------
+
+We propose the following new setup, not repeating any types that remains unchanged::
+
+  -- Primitive type constructors
+  type SORT :: TypeOrConstraint -> RuntimeRep -> Type
+  type IP   :: forall (r :: RuntimeRep). Symbol -> TYPE r -> CONSTRAINT r
+
+  type (=>)  :: forall (r1 :: RuntimeRep) (r2 :: RuntimeRep).
+                CONSTRAINT r1 -> TYPE r2 -> Type  -- primitive
+  type (==>) :: forall (r1 :: RuntimeRep) (r2 :: RuntimeRep).
+                CONSTRAINT r1 -> CONSTRAINT r2 -> Constraint
```

Comment:
I'm not thrilled by the idea of adding another type of arrow to the existing mix, it feels like a lot to keep track of. And `=>` and `==>` feel very similar, in fact, are they not both instantiations of the following?

```
type (=>) :: forall (r1 :: RuntimeRep) (r2 :: RuntimeRep) (tc :: TypeOrConstraint).
             CONSTRAINT r1 -> SORT tc r2 -> SORT tc LiftedRep
```

Of course, this type is even more complex and abstract, and I don't particularly want to expose it to users..

I don't think I have any concrete suggestions here, just registering that the complexity feels bad.

Reply:
Hmm.  We explicitly avoided any polymorphism in `TypeOrConstraint`, in an attempt to avoid complexity.  But you are right that there is a trade-off here: we could have a single, polymorphic type constructor in exchange for, well, it having to be polymorphic.  Indeed we could go further.  We could make FUN do it all
```
type FUN :: forall (m :: Multiplicity) ->
            forall (r1 :: RuntimeRep) (r2 :: RuntimeRep).
            forall (tc1 :: TypeOrConstraint) (tc2 :: TypeOrConstraint)
            SORT tc1 r1 -> SORT tc2 r2 -> Type 
```
These seem like possible futures.  We thought that being fully monomorphic was the most conservative choice -- and one that we could liberalise later -- but I think we'd be open to the committee's guidance on this.  @rae have I missed anything?

Reply:
> I have no idea what polymorphism over TypeOrConstraint would mean. That is, it's effectively polymorphism over visibility... and if a function is polymorphic over visibility, does it take an argument or not?

Hmm, interesting point. I agree that that's a whole can of worms that's best left unopened if we don't have a concrete usecase..

Let me take another approach. 

Here's what struck me about `=>` vs `==>`. If I understand Section 3.1 correctly, then there are two cases where `==>` is required by the type system: instance heads and quantified constraints. Both are syntactically identifiable, evidenced by the fact that we can continue to accept `=>` in those cases for backcompat purposes.

Is the introduction of `==>` as a stable type (see Section 3.4) premature? It seems like we don't actually need to make it public in order to implement the proposal. Suppose instead we leave `==>` in GHC.Prim. We'd be taking away the ability to abstract over `==>` _in an API-stable manner_, but I have a hard time wrapping my head around why you would care to do that in the first place. 

If users come back to us later saying they really do need `==>` as a first-class thing, we can then revisit the question of how polymorphic (or not) to make the various function arrows.

Reply:
@goldfirere also, if we don't know how to reason about polymorphism over `TypeOrConstraint`, we should probably ban it explicitly. I understand that `TypeOrConstraint` will only be exported from `GHC.Prim`, but a user could still write their own polymorphic function over it using non-stable APIs, and it sounds like we wouldn't know what to do with that definition.

Reply:
> Suppose instead we leave ==> in GHC.Prim.

That's an interesting idea.  Indeed we can use it for `(=>)` also.   That is:
* `(=>)` and `(==>)` may indeed exist in Core as `TyCons` but we don't expose them to the user, except maybe through an advertised-as-unsable GHC.Prim -- exactly like `SORT` in fact.
* The *only* places you can write `=>` is in the current, syntactically specified positions; e.g. `f :: Eq a => a -> a`.  For example, you cannot write `T (=>)` or `F ((==>) (Num a))`.

I can't seen a downside to this plan.  It leaves us a bit more wiggle room for the future, and no one is asking for a first class `(=>)` arrow.  @rae?

Maybe `FUN` is the same?  Currently it's exported by `Data.Kind`, but I guess we could change that.  That would be an un-forced change, but one that is consistent with the plan of being more explicit about API stability.   It could mean that some packages would need to change imports -- but how many packages use `FUN` directly?

When I look I see that `GHC.Exts` re-exports all of `GHC.Prim`.  So maybe `GHC.Exts` is advertised-unstable?   For a stable API you want modules without GHC in the module name.

Reply:
> if we don't know how to reason about polymorphism over TypeOrConstraint, we should probably ban it explicitly

There's nothing directly wrong with it.  E.g.
```
data T (a::TypeOrConstraint) where
   T1 :: T TypeLike
   T2 :: T ConstraintLike

f :: forall (a :: TypeOrConstraint). T a -> Int
f T1 = 2
f T2 = 3
```
This is fine; `TypeOrConstraint` is just a vanilla data type.  It's just polymorphic **arrows** that are a problem, and one that we propose to nip in the bud by, well, not providing polymophic arrows.

This does raise another question: maybe `TypeOrConstraint` should be be part of the `GHC.*` exports (unstable) and not the `Data.Kind` exports (stable); see discussion above.

Reply:
I think we currently do a poor job of separating API-stable bits from API-unstable bits. For example, the best way to get `+#` is to import `GHC.Exts`, but `+#` feels very stable to me. On the other hand, we're now proposing some unstable bits.

I could imagine having the unstable bits come from `GHC.Prim` and the stable ones come from `GHC.Exts`. But I don't think we should, for a few reasons:
- The export list of `GHC.Prim` is intimately tied with other implementation choices. We might, for example, take a definition that was previously primitive and make it not so; this would remove the export from `GHC.Prim`, even if we don't actually want to.
- The name `GHC.Prim` does not loudly say that it is unstable. ("Primitive", perhaps, but not unstable.) If someone spots `import GHC.Prim` on some blogpost, they're not going to know that they're building on sand.

I thus think we should make a new place to export unstable bits, perhaps `GHC.Exts.Unstable`. That is, `GHC.Exts` exports GHC-specific (and sometimes magical) things, but we roughly expect it to be stable over time. We might even commit to a migration strategy when making changes. On the other hand, `GHC.Exts.Unstable` can change at any time with no warning. Perhaps even better would be to have a new package `ghc-unstable` exporting unstable modules, and then it would be easier for e.g. an enterprise to scan its dependencies for something untoward. But I'm worried about the possibility of package-import cycles, so it's perhaps best to start small, with `GHC.Exts.Unstable`.

Reply:
As to not exporting things: sure. But I would advocate for exporting `(=>)` from a stable API. That symbol has been around forever, and with the kind (effectively) we are ascribing it. Maybe `(==>)` isn't exported, or exported only from `GHC.Exts.Unstable`. I agree no one is asking for `(==>)`.

I *don't* think we should, at this point, move any definitions around in an unforced manner. That should be done as part of a broader attempt to solidify GHC's stability guarantees.

Reply:
> The name GHC.Prim does not loudly say that it is unstable. 

I agree that changes here are orthogonal to this proposal.  

But (separate to this proposal) my suggestion is that any `import GHC.xxx` signals "building on sand".  Any stable API should not start with "GHC".  That's a simple rule

Reply:
I don't feel strongly about exporting `(=>)`. But anyone can write a type of kind `Constraint -> Type -> Type` if they want to, so it's plausible someone would want to abstract over types of that kind.

I'd be fine with `GHC.xxx` signaling instability, but that goes against a large amount of existing practice. For example, robust libraries (e.g. aeson) are built on `GHC.Generics`, and all primitive operations require imports from the `GHC.` namespace. So I personally think that ship has sailed and we should find a different way of signaling instability.

Reply:
`GHC.Stack` came to my mind as a stable(?) API in GHC..

Reply:
I've always viewed `ghc-prim` (the package) as the unstable API, and the `GHC.*` modules within `base` as relatively stable. Looking again at the breadth of functionality exposed by `base`, that's probably too broad a brush to paint with, but it seems directionally correct to me.

Coincidentally, the proposal suggests exporting the unstable types from `GHC.Prim`, which lives in `ghc-prim`, so I think we're good here. We just need to update the proposal to export `(=>)` and `(==>)` from `GHC.Prim` rather than `Data.Kind`.